VIPs/vip-10.md

```diff
@@ -0,0 +1,109 @@
+---
+vip: 10
+title: Storage node registration (on-chain)
+author: Chris Were <chris@verida.io>
+discussions-to: #dev-protocol on Verida Discord
+status: Draft
+type: Standards Track
+category: Core
+created: 2023-03-22
+---
+
+# Overview
+
+The selection of storage nodes occurs when a DID is first created and when a context is connected to for the first time. The storage nodes are used to store the actual DID Documents and the same nodes are typically (but not necessarily) used to store data for each application context.
+
+This VIP provides a specification for storing the available storage nodes on-chain.
+
+This is phase 2 of [VIP-7: Storage node discovery and selection (off-chain)](./vip-7.md)
+
+This VIP does not implement any token based staking or slashing of nodes, it's purely focused on node addition, removal and discovery.
+
+# Motivation
+
+The current storage node database is centralized and needs to be stored on-chain to decentralize the discovery and selection of nodes on the network.
+
+# Specification
+
+## Data structures
+
+### StorageNode
+
+- didAddress: string
+- endpointUri: string
+- countryCode: string
+- regionCode: string
+- datacenterId: id
+
+### Datacenter
+
+- id: int
+- countryCode: string
+- regionCode: string
+- lat: int
+- long: int
+
+## State variables
+
+TBA
+
+## addNode(didAddress: string, endpointUri: string, countryCode: string, regionCode: string, datacenterId: string, requestSignature: string)
+
+Storage nodes must register themselves with the protocol so they can be discovered by new users. Storage nodes provide additional metadata (via `/status` endpoint) that assists users select the most appropriate storage nodes to use.
+
+It's not possible to re-use a DID to register multiple storage nodes. This is because storage endpoints sign their responses using the DID private key to enhance security of the protocol.
```

Comment:
- A `didAddress` can be used only once for `addNode()` function?
- `endpointUri` is also unique in the contract? There can't be any same `endpointUri` with different `countryCode` or `regionCode`?

Reply:
1. Yes
2. Yes